domain/miningmanager/mempool/orphan_pool.go

```diff
@@ -15,47 +15,102 @@ import (
 	"github.com/pkg/errors"
 )
 
-type idToOrphan map[externalapi.DomainTransactionID]*model.OrphanTransaction
-type previousOutpointToOrphans map[externalapi.DomainOutpoint]idToOrphan
+type idsToOrphans map[externalapi.DomainTransactionID]*model.OrphanTransaction
+type previousOutpointToOrphans map[externalapi.DomainOutpoint]*model.OrphanTransaction
 
 type orphansPool struct {
 	mempool                   *mempool
-	allOrphans                idToOrphan
+	allOrphans                idsToOrphans
 	orphansByPreviousOutpoint previousOutpointToOrphans
 	lastExpireScan            uint64
 }
 
 func newOrphansPool(mp *mempool) *orphansPool {
 	return &orphansPool{
 		mempool:                   mp,
-		allOrphans:                idToOrphan{},
+		allOrphans:                idsToOrphans{},
 		orphansByPreviousOutpoint: previousOutpointToOrphans{},
 		lastExpireScan:            0,
 	}
 }
 
 // this function MUST be called with the mempool mutex locked for writes
 func (op *orphansPool) maybeAddOrphan(transaction *externalapi.DomainTransaction, isHighPriority bool) error {
+	err := op.checkOrphanDuplicate(transaction)
+	if err != nil {
+		return err
+	}
+
+	err = op.checkOrphanSize(transaction)
+	if err != nil {
+		return err
+	}
+	if op.mempool.config.MaximumOrphanTransactionCount <= 0 {
+		return nil
+	}
+	err = op.checkOrphanDoubleSpend(transaction)
+	if err != nil {
+		return err
+	}
+	err = op.limitOrphanPoolSize()
+	if err != nil {
+		return err
+	}
+
+	return op.addOrphan(transaction, isHighPriority)
+}
+
+func (op *orphansPool) limitOrphanPoolSize() error {
+	for len(op.allOrphans) >= op.mempool.config.MaximumOrphanTransactionCount {
+		orphanToRemove := op.randomNonHighPriorityOrphan()
+		if orphanToRemove == nil { // this means all orphans are HighPriority
+			log.Warnf(
+				"Number of high-priority transactions in orphanPool (%d) is higher than maximum allowed (%d)",
+				len(op.allOrphans)+1, // Add + 1 because the current orphan hasn't been added yet
```

Comment:
Yeah, you are right.

This whole complication was because we didn't want a chance to remove the same orphan we just added.
I gave a thought to this, and realized it's not that important, so I just moved limitOrphanPoolSize to the end of maybeAddOrphan (after addOrphan), which made limitOrphanPoolSize much simpler)